slave/process-voms-dirac/lib/process-voms_dirac.pl

```diff
@@ -0,0 +1,345 @@
+			push( @hashes, \%mbr );
+		}
+		else {
+			if ( scalar @components > 3 ) { # Using slower algorithm to match members with commas in their subjects
+				foreach $ca ( @cas ) {
+					$pattern = qr/$ca/;
+					if ( $line =~ /^(.*),${pattern},([^,]*)$/ ) {
+						my %mbr= ( 'DN' => "$1",'CA' => "$ca", 'CN' => getCN($1), 'email' => "$2" );
+						push( @hashes, \%mbr );
+					}
+				}
+			}
+		}
+	}
+	return \@hashes;
+}
+
+
+### effectCall runs actual voms-admin commands. It accepts three arguments:
+#	$command	The shell command to run
+#	$debugMsg	Message to log on execution
+sub effectCall {
+	$command = shift;
+	$debugMsg = shift;
+
+#	printf "$command\n\n";
+	@out=`$command 2>&1`;
+
+	if ( $? == 0 ) {
+		syslog(LOG_INFO, "Done $debugMsg");
+	}
+	else {
+		chomp(@out);
+		syslog(LOG_ERR, "Failed $debugMsg Original message: @out");
+		print STDERR "Failed $debugMsg\nOriginal message: @out\nOriginal command: $command\n";
+		$retval = 1;
+	}
+}
+
+### knownCA indicates whether a given CA is known to the VOMS server. It accepts the user structure
+#	%user		The user whose CA should be checked (DN, CA, email)
+sub knownCA {
+	$ca = $_[0];
+
+	if(grep {$_ eq "$ca"} @cas) {
+		return 1;
+	} else {
+		syslog LOG_ERR, "Unknown CA \"$user->{'CA'}\" requested with user \"$user->{'DN'}\"";
+		print STDERR "Unknown CA \"$user->{'CA'}\" requested with user \"$user->{'DN'}\"\n";
+		return 0;
+	}
+}
+
+
+# This is the actual start.
+
+openlog($program, 'cons,pid', 'user');
+my $retval = 0;
+
+
+# Main parsing loop for the input XML file
+foreach my $vo (@{$vos->{'vo'}}) { # Iterating through individual VOs in the XML
+	$name = $vo->{'name'};
+
+
+	#Collect lists from voms-admin
+	my @groups_current=`voms-admin --vo \Q${name}\E list-groups`;
+	if ( $? != 0 ) {
+		syslog LOG_ERR, "Failed listing groups in VO \"$name\". Error Code $?, original message from voms-admin: @groups_current";
+		print STDERR "Failed listing groups in VO \"$name\". Error Code $?, original message from voms-admin: @groups_current\n";
+		$retval = 1;
+		next;
+	}
+	chomp(@groups_current);
+	s/^\s*// for @groups_current;
+
+	my @roles_current=`voms-admin --vo \Q${name}\E list-roles`;
+	if ( $? != 0 ) {
+		syslog LOG_ERR, "Failed listing roles in VO \"$name\". Error Code $?, original message from voms-admin: @groups_current";
+		print STDERR "Failed listing roles in VO \"$name\". Error Code $?, original message from voms-admin: @groups_current\n";
+		$retval = 1;
+		next;
+	}
+	chomp(@roles_current);
+	s/^\s*Role=// for @roles_current;
+
+	our @cas=`voms-admin --vo \Q${name}\E list-cas`;
+	if ( $? != 0 ) {
+		syslog LOG_ERR, "Failed listing known CAs for VO \"$name\". Error Code $?, original message from voms-admin: @groups_current";
+		print STDERR "Failed listing known CAs for VO \"$name\". Error Code $?, original message from voms-admin: @groups_current\n";
+		$retval = 1;
+		next;
+	}
+	chomp(@cas);
```

Comment:
No, not expected. Of course we are wrapping around a CLI, which can change any moment. But that's another story.